### DISCUSSION TITLE
STM32F030C8T6 GPIO

### DISCUSSION THREAD

**Discussion by ITstreet1** [stm32duino/Arduino_Core_STM32#1671] [Q&A] [answered]
I designed a board around STM32F03C8T6 in the LQFP48 package. When using a STM32CubeProgrammer, selecting UART I can detect an MCU.

```
  10:47:20 : RTS low
  10:47:20 : DTR low
  10:47:20 : Serial Port COM13 is successfully opened.
  10:47:20 : Port configuration: parity = even, baudrate = 115200, data-bit = 8,                     stop-bit = 1.0, flow-control = off
  10:47:20 : Activating device: OK
  10:47:20 : Board       : --
  10:47:20 : Chip ID: 0x440 
  10:47:20 : BootLoader protocol version: 3.1
  10:47:21 : UPLOADING OPTION BYTES DATA ...
  10:47:21 :   Bank          : 0x00
  10:47:21 :   Address       : 0x1ffff800
  10:47:21 :   Size          : 16 Bytes
  10:47:22 : UPLOADING ...
  10:47:22 :   Size          : 1024 Bytes
  10:47:22 :   Address       : 0x8000000
  10:47:22 : Read progress:
  10:47:23 : Data read successfully
  10:47:23 : Time elapsed during the read operation is: 00:00:01.132
  10:47:25 : Disconnected from device.
```
Target info is correct.

I can upload a sketch using a Serial adapter over Arduino IDE. Say Blink sketch.

```
Sketch uses 11784 bytes (17%) of program storage space. Maximum is 65536 bytes.
Global variables use 892 bytes (10%) of dynamic memory, leaving 7300 bytes for local variables. Maximum is 8192 bytes.
      -------------------------------------------------------------------
                       STM32CubeProgrammer v2.9.0                  
      -------------------------------------------------------------------

Serial Port COM13 is successfully opened.
Port configuration: parity = even, baudrate = 115200, data-bit = 8,
                     stop-bit = 1.0, flow-control = off
Activating device: OK
Board       : --
Chip ID: 0x440 
BootLoader protocol version: 3.1
Device name : STM32F05x/F030x8
Flash size  : 64 KBytes (default)
Device type : MCU
Revision ID : --  
Device CPU  : Cortex-M0



Memory Programming ...
Opening and parsing file: Blink.ino.bin
  File          : Blink.ino.bin
  Size          : 12000 Bytes
  Address       : 0x08000000 


Erasing memory corresponding to segment 0:
Erasing internal memory sectors [0 11]
Download in Progress:


File download complete
Time elapsed during download operation: 00:00:02.090

RUNNING Program ... 
  Address:      : 0x8000000
Start operation achieved successfully
```
I designed two buttons, Reset and BOOT, and using them I can switch BOOT modes easily. 

Now, the problem.
No matter what pin I select to make it blink, there is always float output. Of course, I reset the board after uploading a sketch. Here is a schematic:

![Schematic_WHITEBoard White Pill_2022-03-05](https://user-images.githubusercontent.com/30090189/156881495-db45b950-e76a-459e-9093-ed4316144ea3.png)

I double-checked pins on the microscope. Checked with a multimeter. All looks fine. Battery parts are not soldered at all.

**Comment by fpistm** [comment]
Hi,

hard to tell.
If I well understand you can flash properly but not output on any GPIO used to blink?

Maybe you end in error handler which could explains why the pin is in its reset state. Clock issue? Power?

**Reply by ITstreet1** [reply]
If I get an "all ok" message after uploading the sketch, should power, clock, etc, be ok? I will check any suggestions you propose. I can send you links of the components.

I can solder another board, just to eliminate some component issue. But, as I said, on a microscope it looks fine.

**Reply by ITstreet1** [reply]
I double-check the schematic, in case I connect something wrong. 

Now, I am not sure about anything.

**Comment by JiriBilek** [comment]
Just an idea. 
Are you using internal clock or external oscillator in your blink program? I'd suggest to use the internal clock first. 
I suppose the bootloader in ROM uses also the internal oscillator so in case your 8MHz osc is not running you won't notice that until you start your program.

**Comment by ITstreet1** [comment]
![stm32](https://user-images.githubusercontent.com/30090189/157052094-baa0c1e2-bbb9-4f5f-88ca-69a82e359634.jpg)
No option for that. Or do you suggest unsoldering some of the oscilators?

**Reply by ITstreet1** [reply]
The upload method is set to Serial of course.

**Reply by fpistm** [reply]
OK Generic uses the internal clock by default.
I'm not hardware expert. I would test using STM32CubeIDE and debug using STLink. Using STM32CubeMX you should be able to generate a basic project to blink  and see if it stuck and where.

**Reply by fpistm** [reply]
Just to be sure? you flash using STLink SWD (upload method selected on your screenshot) or using Serial?
If you use Serial maybe you have the  issue fixed with #1653

**Reply by fpistm** [reply]
OK. I saw your second comment about Serial usage.
Are you sure you are not in Bootloader mode after the flash ? You should check the Option bytes setting, mainly nBoot1

**Reply by fpistm** [reply]
Just looking the datasheet and it seems you should not use PC13 as OUTPUT to drive a LED:

```
1. PC13, PC14 and PC15 are supplied through the power switch. Since the switch only sinks a limited amount of current
(3 mA), the use of GPIOs PC13 to PC15 in output mode is limited:
- The speed should not exceed 2 MHz with a maximum load of 30 pF.
- These GPIOs must not be used as current sources (e.g. to drive an LED).
```

**Reply by ITstreet1** [reply]
There is no LED soldered. I just measure HIGH and LOW logic with a multimeter. I used any GPIO, the result was the same. No output on any of the GPIOs, and I tried 5-6 of them, random.

With those buttons, I can easily go to Boot mod. While in Boot mode, I can connect to the MCU with STM32CubeProgrammer. When I hit the reset button, I can no longer connect to the chip. So I think it works. When in Boot mode, I can upload sketch using Arduino IDE.

I will try SWD and see if that works.

This about PC13 I didn't know. Glad you point it to me.

**Reply by fpistm** [reply]
Welcome. Maybe you could also try to get support on the forum. I know some people already design their board.

**Reply by ITstreet1** [reply]
https://www.stm32duino.com/
Here?

**Comment by JiriBilek** [comment, marked as answer]
Next, I'd try to disable U(S)ART support in the Arduino IDE. For the blinking sketch you don't need it anyway.

**Comment by ITstreet1** [comment]
LOL. I disabled U(S)ART and it started blinking...
I have not the slightest idea why is that. Although I have to redesign the board for the built-in LED on another pin. I have to make a few tests just to be sure.

**Reply by ITstreet1** [reply]
Ok, I made a few tests, it looks fine. All I had to do is to turn off U(S)ART support.

**Reply by fpistm** [reply]
Anyway this is not normal. Even with U(S)ART support enable this should works.

**Reply by ITstreet1** [reply]
How does the schematic look to you? Maybe I missed something.

**Reply by JiriBilek** [reply]
The schematics looks fine for me. In my experience, the STM's processors are quite tolerant as to the voltage or interference.
I have here a STM32030F6 breakout board, I'll try to reproduce the problem.

**Reply by fpistm** [reply]
As said I'm not an expert. As a first look I would tell it seems OK but can't tell more.
Maybe you can compare with Nucleo schematics.

The only issue I can see if by disabling the U(S)ART it works is that the program use too much RAM leading to error during execution.
I've built the sketch and it seems not the case:
```
Sketch uses 11832 bytes (18%) of program storage space. Maximum is 65536 bytes.
Global variables use 900 bytes (10%) of dynamic memory, leaving 7292 bytes for local variables. Maximum is 8192 bytes.
```
So, I have no more idea about this.
Best way would be to debug to see where it stuck.

**Comment by ITstreet1** [comment]
Can someone change the title? It is STM32F030C8T6

**Reply by fpistm** [reply]
> Can someone change the title? It is STM32F030C8T6

This is already STM32F030C8T6.

**Reply by ITstreet1** [reply]
Check title.

STM32F03C8T6 GPIO

STM32F030C8T6 GPIO

**Reply by fpistm** [reply]
done. Missed the '0'

**Comment by JiriBilek** [comment]
So it's a mystery. I found my old board with STM32F030F6, several times failed to connect the wires correctly (this refers to my statement that STM's survive even bad voltage polarity :) ).
The standard blink sketch works fine for me on PA5.

![image](https://user-images.githubusercontent.com/15474360/157466755-603ae1b8-b6a8-4f7b-aa1f-029d74311d75.png)

**Reply by ITstreet1** [reply]
Not exact MCU. Just saying...

**Reply by JiriBilek** [reply]
I know, I have a shelf full with weird electronics, but there's no STM32F030C unfortunately :-) This is the closest CPU I have.
BTW: If there was a problem in the core that appears in custom boards only, I hope I'll observe it on my board.
FYI my board is the CPU soldered on a breakout board, connected VDD and VDDA, 10k resistor on reset and tantalum capacitor on the power. And that's all. Powered from STLink. I can't imagine more generic hardware.

**Reply by fpistm** [reply]
Have you tried only with one mcu ?

**Reply by JiriBilek** [reply]
Fortunately, I have two CPUs here. I'll try the second one tomorrow but I don't expect a change. Will report.

**Reply by JiriBilek** [reply]
I suspect the question was for the OP not for me. Nevertheless, I tried it with another F030 and unsurprisingly, it worked either.

**Reply by fpistm** [reply]
Yes it is for @ITstreet1

**Reply by ITstreet1** [reply]
No, I didn't. A new revision is on its way. It is not easy to solder it and desolder it multiple times. So I will wait a week or two.

**Comment by ABOSTM** [comment]
I tested my Nucleo F030R8 (closest chipI have in hand).
I used variant Generic F030C8T to be as close as possible to @ITstreet1 setup.
I flashed over UART (PA9/PA10) with boot0 = 3v3 (simple wire)
I used basic/blink example with led built-in on  PA5 (and with default menu: USART enabled generic Serial)
And it is working properly: after setting boot0=0V and pressing Reset button,  led is blinking.

**Reply by ITstreet1** [reply]
Haven't tried SWD yet. This is uploaded with UART.

**Comment by ITstreet1** [comment]
OK.
New PCBs arrived. I soldered everything. I can upload a sketch, and LED blinks on pin PB12. 

Now there are a few issues.
Firstly, the schematic is as above with a changed LED pin. There are RESET and BOOT buttons to enter BOOT mode. I noticed I can upload a sketch no matter am I in a BOOT mode or not. Which is weird. 
Second, after an upload, it works. But if I cut the power and turn it on again it will not start to blink until I press the RESET button. Which is weird, too.

Any idea what to try?

**Reply by ITstreet1** [reply]
While taking a closer look, I noticed on Bluepill C9 cap is 1uF, while I used 100nF. Could it be the issue? R3 resistor is 100K, while I used 1K.

I believe there must be a problem.

https://stm32-base.org/assets/pdf/boards/original-schematic-STM32F103C8T6-Blue_Pill.pdf

**Comment by JiriBilek** [comment]
The capacitor over the reset button is not mandatory, it is used for debouncing the button. The less the value, the less the debouncing effect.

As to the R3, I believe that the value however ridiculously small it is, should not affect the CPU. I have never tried it but I think that you can even short dedicated BOOT pins to GND / VCC.

**Reply by ITstreet1** [reply]
Let's go over the exam again.

While booting with BOOT HIGH, it should execute the program.
While booting with BOOT LOW, it should go into BOOT mode.

This BOOT mode is what I am not sure about. I double-checked the solder joint on a microscope. It looks ok. I tried to upload a code in both modes and it is uploaded BUT NOT stored in flash. It works until I do not disconnect it from the power. So it is not uploaded in BOOT mode. Why this board doesn't go into BOOT mode? How to check is it in BOOT mode or not?

**Comment by ABOSTM** [comment]
@ITstreet1 ,
I disagree with your statement:
Program is stored in flash, and after upload, bootloader jumps directly to your application.
But when you power cycle or press reset button, the CPU runs bootloader code (and this is why your application is not running)
So it looks to me like you always run in bootloader first (and eventually after upload, bootloader jumps to application)
After powercycle, you can try to shortcut BOOT0 to GND (bypassing resistor and capacitor) and press reset button, your application should run.

**Reply by ITstreet1** [reply]
After some intense cleaning with solvents and another check on the microscope, I noticed some of the pins are floating. Not soldered good enough. So I made a few fixes, clean it again, and check the who-knows-what time with a microscope. These are the last conclusions:

-with STM32CubeProgrammer I can see if it is in BOOT mode or not. This part is ok.
-now I can NOT upload a code no matter the BOOT mode with Arduino IDE. Arduino IDE clearly says if it is not in BOOT mode and I can not upload code. If it is in BOOT mode, uploading goes fine. This part is ok.
-after an upload, the board starts executing the program. I don't have to press the RESET button. Is this ok?
-if I upload it, it works. If I press the RESET button while it works, it resets and starts blinking again. This is ok.
-if I power it off and then on, it does not start. I have to press the RESET button. This is not ok.

What to check? What to change?
The upload method is as in the above picture.

**Comment by ABOSTM** [comment]
> -after an upload, the board starts executing the program. I don't have to press the RESET button. Is this ok?

Yes it is ok: at the end of upload, bootloader jumps automatically to application

> -if I power it off and then on, it does not start. I have to press the RESET button. This is not ok.

I am not that expert in analog, but my interpretation of this phenomenon would reside in transient level of BOOT0:
When you power up the board the voltage of BOOT0 capacitor will start from 0 and rise to 3,3V. During this rising phase, capacitor will be charging thank to current flowing through Boot0 resistor.
In other words, BOOT0 voltage will start with non 0 value (3,3V ?) and will decrease to 0V.

So what is the sate of BOOT0 when it is evaluated to determined whether to run application or boot loader ? I don't know.

What you can do, for test purpose, is to solder/wire BOOT0 to GND directly, and check if application runs after powerup.

**Reply by JiriBilek** [reply]
Good point with the capacitor on BOOT. It has no business being there, no debouncing needed.

**Reply by ITstreet1** [reply]
You guys rock. C15 out, problems out.

Can someone tell me a word or two about this:
https://github.com/stm32duino/Arduino_Core_STM32/discussions/1671#discussioncomment-2318824
I can upload a sketch with or without U(S)ART support, but it will work only if it is disabled.

Can I use the Serial terminal just as on Uno? Tried AnalogReadSerial, but get an error while compiling.

Any other thing I should check?

I will try some OLED I2C and will report here what I get.

**Reply by ITstreet1** [reply]
PB6 and PB7 should be SCL and SDA for I2C lines. 
https://datasheet.lcsc.com/lcsc/1811061717_STMicroelectronics-STM32F030C8T6_C23922.pdf

Tried U8g2 and Adafruit SSD1306 libs, but no luck.
For Adafruit it trows Serial not declared or so. Could be an issue with disabled USART.
For U8g2, I tried a few from the list for a simple Hello World example. Nothing.

**Reply by fpistm** [reply]
You have to use a resistor pull up on each lines (typically 4.7K).

**Reply by fpistm** [reply]
> I can upload a sketch with or without U(S)ART support, but it will work only if it is disabled.
This is not normal. Even if U(S)ART support is enabled as it is not used it could not interfere with the blink.


> Can I use the Serial terminal just as on Uno? Tried AnalogReadSerial, but get an error while compiling.

Yes the core is fully Arduino API compatible.  When you told  `AnalogReadSerial`, I guess you talk about the default example. It works.

```
// the setup routine runs once when you press reset:
void setup() {
  // initialize serial communication at 9600 bits per second:
  Serial.begin(9600);
}

// the loop routine runs over and over again forever:
void loop() {
  // read the input on analog pin 0:
  int sensorValue = analogRead(A0);
  // print out the value you read:
  Serial.println(sensorValue);
  delay(1);        // delay in between reads for stability
}
```
In your case for generic STM32F030C8Tx, A0 is PA0:
https://github.com/stm32duino/Arduino_Core_STM32/blob/6588dee03382e73ed42c4a5e473900ab3b79d6e4/variants/STM32F0xx/F030C8T/variant_generic.h#L18

And of course U(S)ART support have to be enabled. using this by default:
https://github.com/stm32duino/Arduino_Core_STM32/blob/6588dee03382e73ed42c4a5e473900ab3b79d6e4/variants/STM32F0xx/F030C8T/variant_generic.h#L123-L135

Honestly based on your input it seems your environment is not correct. Maybe install has been corrupted?

**Reply by ITstreet1** [reply]
For UART
Code will run only when I upload it with disabled Serial. With disabled Serial, code will not compile. With enabled Serial, code will compile and upload, but it will not run.

For I2C
I can not upload the Adafruit example because it uses so many Serial prints. It is all over lib, in Busio too.
U8g2 doesn't work. At least I can not find what instance to use for it.

This project is nothing important. I just have a few of these MCUs, so why not make something with them, and learn along the way.
Maybe someone will need this in the future. :)

**Reply by ITstreet1** [reply]
BOOM!!!!

Oled works with U8g2.
Pins are PB6 for SCL and PB7 for SDA.
Instance in the lib:
U8G2_SSD1306_128X64_NONAME_F_SW_I2C u8g2(U8G2_R0, /* clock=*/ PB6, /* data=*/ PB7, /* reset=*/ U8X8_PIN_NONE);
The screen already has pull-up resistors, so I didn't need them.

USART disabled, of course.

Must try something with SPI.

**Comment by fpistm** [comment]
Could you share the build errors?

**Reply by ITstreet1** [reply]
I get errors when building with disabled USART. Which is logical. With enabled USART code will not execute. 
If you think for those with disabled USART, it just says Serial is not defined.

**Reply by fpistm** [reply]
OK. Anyway I've tested and it works as expected.🤨

**Reply by ITstreet1** [reply]
No idea why I have to disable USART?
Maybe to post this on https://www.stm32duino.com/?

**Reply by fpistm** [reply]
You can try.  Maybe some users will have some more idea.
At this stage, I guess only debug would help to understand why when the U(S)ART is enabled it does not work.

When we met this, it is due a small mcu with not enough flash so enabling it consume much space and explains why it does not run.
Are you sure about the mcu reference?

**Comment by ITstreet1** [comment]
![Schematic](https://user-images.githubusercontent.com/30090189/162228536-d93e9a3a-8490-4797-b906-a86d70ca4655.png)

Sketch uses 8044 bytes (12%) of program storage space. Maximum is 65536 bytes.
Global variables use 552 bytes (6%) of dynamic memory, leaving 7640 bytes for local variables. Maximum is 8192 bytes.
      -------------------------------------------------------------------
                       STM32CubeProgrammer v2.9.0                  
      -------------------------------------------------------------------

Serial Port COM13 is successfully opened.
Port configuration: parity = even, baudrate = 115200, data-bit = 8,
                     stop-bit = 1.0, flow-control = off
Activating device: OK
Board       : --
Chip ID: 0x440 
BootLoader protocol version: 3.1
Device name : STM32F05x/F030x8
Flash size  : 64 KBytes (default)
Device type : MCU
Revision ID : --  
Device CPU  : Cortex-M0



Memory Programming ...
Opening and parsing file: Blink.ino.bin
  File          : Blink.ino.bin
  Size          : 8256 Bytes
  Address       : 0x08000000 


Erasing memory corresponding to segment 0:
Erasing internal memory sectors [0 8]
Download in Progress:


File download complete
Time elapsed during download operation: 00:00:01.448

RUNNING Program ... 
  Address:      : 0x8000000
Start operation achieved successfully

**Reply by ITstreet1** [reply]
This blinks the LED.

**Reply by ITstreet1** [reply]
Sketch uses 11784 bytes (17%) of program storage space. Maximum is 65536 bytes.
Global variables use 892 bytes (10%) of dynamic memory, leaving 7300 bytes for local variables. Maximum is 8192 bytes.
      -------------------------------------------------------------------
                       STM32CubeProgrammer v2.9.0                  
      -------------------------------------------------------------------

Serial Port COM13 is successfully opened.
Port configuration: parity = even, baudrate = 115200, data-bit = 8,
                     stop-bit = 1.0, flow-control = off
Activating device: OK
Board       : --
Chip ID: 0x440 
BootLoader protocol version: 3.1
Device name : STM32F05x/F030x8
Flash size  : 64 KBytes (default)
Device type : MCU
Revision ID : --  
Device CPU  : Cortex-M0



Memory Programming ...
Opening and parsing file: Blink.ino.bin
  File          : Blink.ino.bin
  Size          : 12000 Bytes
  Address       : 0x08000000 


Erasing memory corresponding to segment 0:
Erasing internal memory sectors [0 11]
Download in Progress:


File download complete
Time elapsed during download operation: 00:00:02.052

RUNNING Program ... 
  Address:      : 0x8000000
Start operation achieved successfully

**Reply by ITstreet1** [reply]
This is when I turn on USART support: Generic Serial
